File: ios/Models/SipConfiguration.swift
//
//  SipConfiguration.swift
//  Voip24hSdk
//
//  Created by Phát Nguyễn on 27/11/2023.
//  Copyright © 2023 Facebook. All rights reserved.
//

import Foundation
import linphonesw

struct SipConfiguration: Codable {
    var ext: String = ""
    var password: String = ""
    var domain: String = ""
    var port: Int = 5060
    var transport: String = ""
    var isKeepAlive: Bool = false
    
    init(dictionary: [String: Any]) throws {
        self = try JSONDecoder().decode(SipConfiguration.self, from: JSONSerialization.data(withJSONObject: dictionary))
    }
    
    private enum CodingKeys: String, CodingKey {
        case ext = "extension"
        case password, domain, port, transport, isKeepAlive
    }
    
    func toLpTransportType() -> TransportType {
        switch(transport) {
            case "Udp":
                return TransportType.Udp
            case "Tcp":
                return TransportType.Tcp
            case "Tls":
                return TransportType.Tls
            case "Dtls":
                return TransportType.Dtls
            default:
                return TransportType.Udp
        }
    }
}

File: src/model/SipConfiguration.js
import { TransportType } from "../enum_type/EnumType";

class SipConfiguration {
    constructor(ext = "", pass = "", domain = "", port, isKeepAlive, transport) {
        this.extension = ext;
        this.password = pass;
        this.domain = domain;
        this.port = port;
        this.isKeepAlive = isKeepAlive;
        this.transport = transport;
    }
}

export default class SipConfigurationBuilder {

    port: int = 5060;
    transport: TransportType = TransportType.Udp;
    isKeepAlive: bool = false;

    constructor(extension = "", password = "", domain = "") {
        this.extension = extension;
        this.password = password;
        this.domain = domain;
    }

    setPort(port) {
        this.port = port;
        return this;
    }

    setTransportType(transportType) {
        this.transport = transportType;
        return this;
    }

    setKeepAlive(isKeepAlive) {
        this.isKeepAlive = isKeepAlive;
        return this;
    }

    build() {
        var sipConfiguration = new SipConfiguration(
            this.extension, 
            this.password, 
            this.domain, 
            this.port, 
            this.isKeepAlive, 
            this.transport
        );
        return sipConfiguration;
    }
}